source/Tree/Tree.js

```diff
@@ -0,0 +1,439 @@
+
+  /** Width of list */
+  width: number,
+};
+
+/**
+ * This component renders a tree of elements using powerful virtualization technology to render
+ * only visible elements to the HTML document.
+ */
+export default class Tree extends React.PureComponent {
+  static defaultProps = {
+    autoHeight: false,
+    estimatedRowSize: 30,
+    nodeNestingMultiplier: 10,
+    noRowsRenderer: () => null,
+    onRowsRendered: () => {},
+    onScroll: () => {},
+    overscanIndicesGetter: accessibilityOverscanIndicesGetter,
+    overscanRowCount: 10,
+    rowRenderer: defaultRowRenderer,
+    scrollToAlignment: 'auto',
+    scrollToIndex: -1,
+    style: {},
+  };
+
+  props: Props;
+  _grid: ?Grid;
+  _nodes: Node[] = [];
+  _nodesStates: { [id: string]: boolean } = {};
+
+  forceUpdateGrid() {
+    if (this._grid) {
+      this._grid.forceUpdate();
+    }
+  }
+
+  /**
+   * Sets nodes states (opened/closed) to a states defined in `boolOrObject` parameter.
+   * @param states object that contains nodes' ids as keys and boolean states as values.
+   */
+  setNodesStates(states: { [id: string]: boolean }) {
+    this._nodesStates = {...this._nodesStates, ...states};
+    this.forceUpdate();
+  }
+
+  /** See Grid#getOffsetForCell */
+  getOffsetForRow({alignment, index}: { alignment: Alignment, index: number }) {
+    if (this._grid) {
+      const {scrollTop} = this._grid.getOffsetForCell({
+        alignment,
+        columnIndex: 0,
+        rowIndex: index,
+      });
+
+      return scrollTop;
+    }
+    return 0;
+  }
+
+  /** CellMeasurer compatibility */
+  invalidateCellSizeAfterRender({columnIndex, rowIndex}: CellPosition) {
+    if (this._grid) {
+      this._grid.invalidateCellSizeAfterRender({
+        columnIndex,
+        rowIndex,
+      });
+    }
+  }
+
+  /** See Grid#measureAllCells */
+  measureAllRows() {
+    if (this._grid) {
+      this._grid.measureAllCells();
+    }
+  }
+
+  /** CellMeasurer compatibility */
+  recomputeGridSize({columnIndex = 0, rowIndex = 0}: CellPosition = {}) {
+    if (this._grid) {
+      this._grid.recomputeGridSize({
+        columnIndex,
+        rowIndex,
+      });
+    }
+  }
+
+  /** See Grid#recomputeGridSize */
+  recomputeRowHeights(index: number = 0) {
+    if (this._grid) {
+      this._grid.recomputeGridSize({
+        columnIndex: 0,
+        rowIndex: index,
+      });
+    }
+  }
+
+  /** See Grid#scrollToPosition */
+  scrollToPosition(scrollTop: number = 0) {
+    if (this._grid) {
+      this._grid.scrollToPosition({scrollTop});
+    }
+  }
+
+  /** See Grid#scrollToCell */
+  scrollToRow(index: number = 0) {
+    if (this._grid) {
+      this._grid.scrollToCell({
+        columnIndex: 0,
+        rowIndex: index,
+      });
+    }
+  }
+
+  render() {
+    const {
+      className,
+      noRowsRenderer,
+      scrollToIndex,
+      width,
+    } = this.props;
+
+    const classNames = cn('ReactVirtualized__Tree', className);
+
+    const rowCount = this._prepareNodes();
```

Comment:
Creating an array with thousands of objects seems like a potentially expensive thing to compute each `render` (eg any time a `Tree` prop like `width` or `height` changes, any time a node is expanded/collapsed).

I'm not super familiar with this component, but shouldn't it be possible to move this pre-compute step into `componentWillMount` and `componentWillReceiveProps` (where we could only recompute it if it seems likely to have changed- eg not in response to a changed `width` or something)?

Would it be possible to only apply a patch to the computed nodes when a single node is collapsed/expanded (rather than recomputing the entire tree)?

Reply:
Yes, creating an internal array was a problem I've stuck from the very start. Thanks for a fresh idea, I'll think about it.